iothub/device/src/Transport/RetryDelegatingHandler.cs

```diff
@@ -226,6 +227,90 @@ public override async Task<Message> ReceiveAsync(TimeoutHelper timeoutHelper)
             }
         }
 
+        public override async Task EnableReceiveMessageAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (Logging.IsEnabled)
+                {
+                    Logging.Enter(this, cancellationToken, nameof(EnableReceiveMessageAsync));
+                }
+
+                await _internalRetryPolicy
+                    .ExecuteAsync(
+                        async () =>
+                        {
+                            // Ensure that the connection has been opened, before enabling the callback for receiving messages.
+                            await EnsureOpenedAsync(cancellationToken).ConfigureAwait(false);
+
+                            // Wait to acquire the _handlerLock. This ensures that concurrently invoked API calls are invoked in a thread-safe manner.
+                            await _handlerLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+                            try
+                            {
+                                // The telemetry downlink needs to be enabled only for the first time that the callback is set.
+                                Debug.Assert(!_deviceReceiveMessageEnabled);
+                                await base.EnableReceiveMessageAsync(cancellationToken).ConfigureAwait(false);
+                                _deviceReceiveMessageEnabled = true;
+                            }
+                            finally
+                            {
+                                _handlerLock.Release();
```

Comment:
could this throw a SemaphoreFullException?

Reply:
no, it shouldn't. We grab the semaphore outside of the try block, so we wouldn't hit this finally block without having grabbed the semaphore first.